### PR TITLE
fix: slot and primary can be different shape

### PR DESCRIPTION
**Pull request by candyzone**
See github issue #19457

### CODE REVIEW

**Comment by candyzone** [comment]
@allenlavoie 
the pr “https://github.com/tensorflow/tensorflow/pull/20539” relateed commit is wrong, i create a new pr again.

File: tensorflow/python/training/slot_creator.py [hunk 1]
```diff
@@ -76,11 +76,16 @@ def _create_slot_var(primary, val, scope, validate_shape, shape, dtype):
...
-    slot._set_save_slice_info(variables.Variable.SaveSliceInfo(
-        slice_info.full_name + "/" + real_slot_name,
-        slice_info.full_shape[:],
-        slice_info.var_offset[:],
-        slice_info.var_shape[:]))
+    # supprot slot's shape not same as primary's shape
+    # example: primary's shape = [10, 20, 30], slot's shape =
+    # [], [10], [10, 20] or [10, 20, 30] is allowed
+    n = slot.shape.ndims
+    if (n > 0):
```
Comment (allenlavoie) [inline tensorflow/python/training/slot_creator.py]: Why is this condition necessary? Seems like scalars should follow the same pattern.

File: tensorflow/python/training/slot_creator.py [hunk 2]
```diff
@@ -76,11 +76,16 @@ def _create_slot_var(primary, val, scope, validate_shape, shape, dtype):
...
+      slot._set_save_slice_info(variables.Variable.SaveSliceInfo(
+          slice_info.full_name + "/" + real_slot_name,
+          slice_info.full_shape[:n],
```
Comment (allenlavoie) [inline tensorflow/python/training/slot_creator.py]: Isn't the full shape always just slot.shape? Why slice?

File: tensorflow/python/training/slot_creator.py (hunk 1 above)
Comment (candyzone) [reply tensorflow/python/training/slot_creator.py]: detail info. 
https://github.com/tensorflow/tensorflow/issues/19457
Some optimizer creates slot that should not has the same shape with primary.

File: tensorflow/python/training/slot_creator.py (hunk 2 above)
Comment (candyzone) [reply tensorflow/python/training/slot_creator.py]: A slot is a `Variable` created with the same first m-dimension as a primary `Tensor`. A slot is always scoped in the namespace of the primary object and	variable or `Tensor`
So, slot's slice info can be slice of the primary's slice info

File: tensorflow/python/training/slot_creator.py (hunk 1 above)
Comment (allenlavoie) [reply tensorflow/python/training/slot_creator.py]: That doesn't really answer the question. Why are scalars (n == 0) treated differently than other tensors (n > 0)?

File: tensorflow/python/training/slot_creator.py (hunk 2 above)
Comment (allenlavoie) [reply tensorflow/python/training/slot_creator.py]: Yes, but the full shape is just "slot.shape" always. There's no need to slice.

Comment (candyzone) [reply tensorflow/python/training/slot_creator.py]: the 'slice_info.full_shape' is the primary's slice_info, not the slot's slice_info.
slot's slice_info may be different from primary's slice_info.

File: tensorflow/python/training/slot_creator.py (hunk 1 above)
Comment (candyzone) [reply tensorflow/python/training/slot_creator.py]: if n==0, slot cannot slice, every slot with primary should duplicate. if n>0, slot can be slice, every slot with primary can be slice

Comment (allenlavoie) [reply tensorflow/python/training/slot_creator.py]: `lst[:0]` works fine and returns an empty list. I still don't understand the issue. If there is one, could you add a unit test which fails without the condition?

File: tensorflow/python/training/slot_creator.py (hunk 2 above)
Comment (allenlavoie) [reply tensorflow/python/training/slot_creator.py]: Right, `slice_info.full_shape` would need to be sliced. But `slot.shape` wouldn't. This is nitpicky, so I'm fine leaving it slicing. But do you agree that `slot.shape == slice_info.full_shape[:n]`? If not, a unit test where it fails would help me understand.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
